File: lintcode/python/0661_convert_bst_to_greater_tree.py
#!/usr/bin/python -t

# dfs, traversal way
#从右子树开始递归，每次加上右子树的和，再递归左子树即可。

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of binary tree
    @return: the new root
    """
    def convertBST(self, root):
        # write your code here
        self.sum = 0
        self.dfs(root)
        return root
        
    def dfs(self, node):
        if node == None:
            return
        
        if node.right:
            self.dfs(node.right)
            
        self.sum += node.val
        node.val = self.sum
        
        if node.left:
            self.dfs(node.left)
            

File: lintcode/golang/0661_convert_bst_to_greater_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the root of binary tree
 * @return: the new root
 */

var val int 

func convertBST (root *TreeNode) *TreeNode {
    // write your code here
    dfs(root)
    return root
}

func dfs (node *TreeNode) {
    if node == nil {
        return
    }
    
    dfs(node.Right)
    val += node.Val
    node.Val = val
    dfs(node.Left)
    
    return
}
